File: src/main/java/org/taktik/icure/logic/impl/InvoiceLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.google.common.base.Strings;
import org.ektorp.ComplexKey;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.InvoiceDAO;
import org.taktik.icure.dao.Option;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.dto.data.LabelledOccurence;
import org.taktik.icure.entities.Form;
import org.taktik.icure.entities.Insurance;
import org.taktik.icure.entities.Invoice;
import org.taktik.icure.entities.Patient;
import org.taktik.icure.entities.User;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.entities.embed.InvoiceType;
import org.taktik.icure.entities.embed.InvoicingCode;
import org.taktik.icure.entities.embed.MediumType;
import org.taktik.icure.exceptions.DeletionException;
import org.taktik.icure.logic.InvoiceLogic;
import org.taktik.icure.logic.UserLogic;
import org.taktik.icure.utils.FuzzyValues;

@Service
public class InvoiceLogicImpl extends GenericLogicImpl<Invoice, InvoiceDAO> implements InvoiceLogic {
	private static final Logger log = LoggerFactory.getLogger(InvoiceLogicImpl.class);

	private UserLogic userLogic;
	private InvoiceDAO invoiceDAO;

	@Override
	public Invoice createInvoice(Invoice invoice) {
		return invoiceDAO.create(invoice);
	}

	@Override
	public String deleteInvoice(String invoiceId) throws DeletionException {
		try {
			deleteEntities(Arrays.asList(invoiceId));
			return invoiceId;
		} catch (Exception e) {
			throw new DeletionException(e.getMessage(), e);
		}
	}

	@Override
	public Invoice getInvoice(String invoiceId) {
		return invoiceDAO.get(invoiceId);
	}

	@Override
	public List<Invoice> getInvoices(List<String> ids) {
		return invoiceDAO.getList(ids);
	}

	@Override
	public Invoice modifyInvoice(Invoice invoice) {
		return invoiceDAO.save(invoice);
	}

	@Override
	public List<Invoice> updateInvoices(List<Invoice> invoices) {
		return invoiceDAO.save(invoices);
	}

	@Override
	public Invoice addDelegation(String invoiceId, Delegation delegation) {
		Invoice invoice = invoiceDAO.get(invoiceId);
		invoice.addDelegation(delegation.getDelegatedTo(), delegation);
		return invoiceDAO.save(invoice);
	}

	@Override
	public PaginatedList<Invoice> findByAuthor(String hcParty, String userId, Long fromDate, Long toDate, PaginationOffset paginationOffset) {
		return invoiceDAO.findByHcParty(hcParty, fromDate, toDate, paginationOffset);
	}

	@Override
	public List<Invoice> listByHcPartyContacts(String hcParty, Set<String> contactIds) {
		return invoiceDAO.listByHcPartyContacts(hcParty, contactIds);
	}

	@Override
	public List<Invoice> listByHcPartyRecipientIds(String hcParty, Set<String> recipientIds) {
		return invoiceDAO.listByHcPartyRecipientIds(hcParty, recipientIds);
	}

	@Override
	public List<Invoice> listByHcPartyPatientSks(String hcParty, Set<String> secretPatientKeys) {
		return invoiceDAO.listByHcPartyPatientFk(hcParty, secretPatientKeys);
	}

	@Override
	public List<Invoice> listByHcPartySentMediumTypeInvoiceTypeSentDate(String hcParty, MediumType sentMediumType, InvoiceType invoiceType, boolean sent, Long fromDate, Long toDate) {
		return invoiceDAO.listByHcPartySentMediumTypeInvoiceTypeSentDate(hcParty, sentMediumType, invoiceType, sent, fromDate, toDate);
	}

	@Override
	public List<Invoice> listByHcPartySendingModeStatus(String hcParty, String sendingMode, String status, Long fromDate, Long toDate) {
		return invoiceDAO.listByHcPartySendingModeStatus(hcParty, sendingMode, status, fromDate, toDate);
	}

	@Override
    public List<Invoice> listByHcPartyGroupId(String hcParty, String groupId) {
        return invoiceDAO.listByHcPartyGroupId(hcParty, groupId);
    }

    @Override
	public List<Invoice> listByHcPartyRecipientIdsUnsent(String hcParty, Set<String> recipientIds) {
		return invoiceDAO.listByHcPartyRecipientIdsUnsent(hcParty, recipientIds);
	}

	@Override
	public List<Invoice> listByHcPartyPatientSksUnsent(String hcParty, Set<String> secretPatientKeys) {
		return invoiceDAO.listByHcPartyPatientFkUnsent(hcParty, secretPatientKeys);
	}

	@Override
	public List<Invoice> listByServiceIds(Set<String> serviceIds) {
		return invoiceDAO.listByServiceIds(serviceIds);
	}


	@Override
	public Invoice mergeInvoices(String hcParty, List<Invoice> invoices, Invoice destination) throws DeletionException {
		if (destination.getInvoicingCodes() == null) { destination.setInvoicingCodes(new ArrayList<>()); }
		for (Invoice i:invoices) {
			destination.getInvoicingCodes().addAll(i.getInvoicingCodes());
			deleteInvoice(i.getId());
		}

		return modifyInvoice(destination);
	}

	@Override
	public Invoice validateInvoice(String hcParty, Invoice invoice, String refScheme, String forcedValue) {
		if (forcedValue != null || !Strings.isNullOrEmpty(invoice.getInvoiceReference())) {
			invoice.setInvoiceReference(forcedValue);
		} else {
			if (refScheme == null) {
				refScheme = "yyyy00000";
			}
			LocalDateTime ldt = LocalDateTime.ofInstant(Instant.ofEpochMilli(invoice.getInvoiceDate()), ZoneId.systemDefault());

			NumberFormat f = new DecimalFormat("00");

			String startScheme = refScheme.replaceAll("yyyy", "" + ldt.getYear()).replaceAll("MM", f.format(ldt.getMonthValue())).replaceAll("dd", "" + f.format(ldt.getDayOfMonth()));
			String endScheme = refScheme.replaceAll("0", "9").replaceAll("yyyy", "" + ldt.getYear()).replaceAll("MM", "" + f.format(ldt.getMonthValue())).replaceAll("dd", "" + f.format(ldt.getDayOfMonth()));

			List<Invoice> prevInvoices = invoiceDAO.listByHcPartyReferences(hcParty, endScheme, null, true, 1);
			invoice.setInvoiceReference("" + (prevInvoices.size() > 0 && prevInvoices.get(0).getInvoiceReference() != null ? Math.max(Long.valueOf(prevInvoices.get(0).getInvoiceReference()) + 1L,Long.valueOf(startScheme) + 1L) : Long.valueOf(startScheme) + 1L));
		}
		invoice.setSentDate(System.currentTimeMillis());
		return modifyInvoice(invoice);
	}

	@Override
	public List<Invoice> appendCodes(String hcPartyId, String userId, String insuranceId, Set<String> secretPatientKeys, InvoiceType type, MediumType sentMediumType, List<InvoicingCode> invoicingCodes, String invoiceId, Integer invoiceGraceTime) {
		if (sentMediumType == MediumType.efact) {
			invoicingCodes.forEach(c->c.setPending(true));
		}
		final int invoiceGraceTimeInDays = invoiceGraceTime == null ? 0 : invoiceGraceTime;
		Invoice selectedInvoice = (invoiceId != null) ? this.getInvoice(invoiceId) : null;

		List<Invoice> invoices = selectedInvoice != null ? new ArrayList<>() : this.listByHcPartyPatientSksUnsent(hcPartyId,secretPatientKeys).stream().filter(i->
				i.getInvoiceType() == type && i.getSentMediumType() == sentMediumType && (insuranceId == null ? i.getRecipientId() == null  : insuranceId.equals(i.getRecipientId()))
		).collect(Collectors.toList());

		if (selectedInvoice == null &&  invoices.isEmpty()) {
			invoices = this.listByHcPartyRecipientIdsUnsent(hcPartyId,Collections.singleton(insuranceId)).stream().filter(i->
					i.getInvoiceType() == type && i.getSentMediumType() == sentMediumType && i.getSecretForeignKeys().equals(secretPatientKeys)
			).collect(Collectors.toList());
		}

		Set<Invoice> modifiedInvoices = new HashSet<>();
		Set<Invoice> createdInvoices = new HashSet<>();

		for (InvoicingCode invoicingCode : new ArrayList<>(invoicingCodes)) {
			LocalDateTime icDateTime = FuzzyValues.getDateTime(invoicingCode.getDateCode());

			Optional<Invoice> unsentInvoice = selectedInvoice != null ? Optional.of(selectedInvoice) : invoices.stream().filter(i ->
					i.getInvoiceDate() != null && Math.abs(FuzzyValues.getDateTime(i.getInvoiceDate()).until(icDateTime, ChronoUnit.DAYS)) <= invoiceGraceTimeInDays
			).findAny();

			if (unsentInvoice.isPresent()) {
				Invoice invoice = unsentInvoice.get();
				invoice.getInvoicingCodes().add(invoicingCode);

				if (!createdInvoices.contains(invoice)) { modifiedInvoices.add(modifyInvoice(invoice)); }
			} else {
				Invoice newInvoice = new Invoice();
				newInvoice.setInvoiceDate(invoicingCode.getDateCode()!=null?invoicingCode.getDateCode():System.currentTimeMillis());
				newInvoice.setInvoiceType(type);
				newInvoice.setSentMediumType(sentMediumType);
				newInvoice.setRecipientId(insuranceId);
				newInvoice.setRecipientType((type == InvoiceType.mutualfund || type == InvoiceType.payingagency) ? Insurance.class.getName() : Patient.class.getName());
				newInvoice.setInvoicingCodes(invoicingCodes);
				newInvoice.setAuthor(userId);
				newInvoice.setResponsible(hcPartyId);
				newInvoice.setCreated(System.currentTimeMillis());
				newInvoice.setModified(newInvoice.getCreated());

				//The invoice must be completed with ids and delegations and created on the server
				createdInvoices.add(newInvoice);
				invoices.add(newInvoice);
			}
		}
		ArrayList<Invoice> result = new ArrayList<>(createdInvoices);

		result.addAll(modifiedInvoices);
		return result;
	}

	@Override
	public Invoice addDelegations(String invoiceId, List<Delegation> delegations) {
		Invoice invoice = invoiceDAO.get(invoiceId);
		delegations.forEach(d->invoice.addDelegation(d.getDelegatedTo(),d));
		return invoiceDAO.save(invoice);
	}

	@Override
	public List<Invoice> removeCodes(String userId, Set<String> secretPatientKeys, String serviceId, List<String> tarificationIds) {
		User user = this.userLogic.getUser(userId);
		List<Invoice> invoices = this.listByHcPartyPatientSksUnsent(user.getHealthcarePartyId(), secretPatientKeys).stream()
				.filter(i -> i.getInvoicingCodes().stream().anyMatch(ic -> serviceId.equals(ic.getServiceId()) && tarificationIds.contains(ic.getTarificationId()))
		).sorted((a,b)-> ((Long)(b.getInvoiceDate()!=null?b.getInvoiceDate():99999999999999L)).compareTo(a.getInvoiceDate())).collect(Collectors.toList());
		List<Invoice> result = new LinkedList<>();
		for (Invoice i:invoices) {
			boolean hasChanged = false;
			List<InvoicingCode> l = new LinkedList<>(i.getInvoicingCodes());
			for(InvoicingCode ic:i.getInvoicingCodes()) {
				if (tarificationIds.contains(ic.getTarificationId())) {
					l.remove(ic);
					tarificationIds.remove(ic.getTarificationId());
					hasChanged = true;
				}
			}
			if (hasChanged) {
				i.setInvoicingCodes(l);
				result.add(modifyInvoice(i));
			}
		}
		return result;
	}

	@Override
	public List<Invoice> listAllHcpsByStatus(String status, Long from, Long to, List<String> hcpIds) {
		if (status == null) {throw new IllegalArgumentException("Status cannot be null"); }
		return invoiceDAO.listAllHcpsByStatus(status,from,to,hcpIds);
	}

	@Override
	public void solveConflicts() {
		List<Invoice> invoicesInConflict = invoiceDAO.listConflicts().stream().map(it -> invoiceDAO.get(it.getId(), Option.CONFLICTS)).collect(Collectors.toList());

		invoicesInConflict.forEach(iv -> {
			Arrays.stream(iv.getConflicts()).map(c -> invoiceDAO.get(iv.getId(), c)).forEach(cp -> {
				iv.solveConflictWith(cp);
				invoiceDAO.purge(cp);
			});
			invoiceDAO.save(iv);
		});
	}

	@Override
	public List<LabelledOccurence> getTarificationsCodesOccurences(String hcPartyId, long minOccurences) {
		return invoiceDAO.listTarificationsFrequencies(hcPartyId).parallelStream()
				.filter(v -> v.getValue() != null && v.getValue() >= minOccurences)
				.map(v -> new LabelledOccurence((String) v.getKey().getComponents().get(1), v.getValue()))
				.sorted(Comparator.comparing(LabelledOccurence::getOccurence).reversed())
				.collect(Collectors.toList());
	}

	@Autowired
	public void setInvoiceDAO(InvoiceDAO invoiceDAO) {
		this.invoiceDAO = invoiceDAO;
	}

	@Autowired
	public void setUserLogic(UserLogic userLogic) {
		this.userLogic = userLogic;
	}

	@Override
	protected InvoiceDAO getGenericDAO() {
		return invoiceDAO;
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/InvoiceLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import com.google.common.base.Strings
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.couchdb.entity.Option
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.icure.asyncdao.InvoiceDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.EntityReferenceLogic
import org.taktik.icure.asynclogic.InvoiceLogic
import org.taktik.icure.asynclogic.UserLogic
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.domain.filter.chain.FilterChain
import org.taktik.icure.dto.data.LabelledOccurence
import org.taktik.icure.entities.Insurance
import org.taktik.icure.entities.Invoice
import org.taktik.icure.entities.Patient
import org.taktik.icure.entities.embed.Delegation
import org.taktik.icure.entities.embed.InvoiceType
import org.taktik.icure.entities.embed.InvoicingCode
import org.taktik.icure.entities.embed.MediumType
import org.taktik.icure.exceptions.DeletionException
import org.taktik.icure.utils.FuzzyValues
import org.taktik.icure.utils.toComplexKeyPaginationOffset
import java.text.DecimalFormat
import java.text.NumberFormat
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.*
import kotlin.math.abs
import kotlin.math.max

@ExperimentalCoroutinesApi
@Service
class InvoiceLogicImpl(private val filters: Filters,
                       private val userLogic: UserLogic,
                       private val uuidGenerator: UUIDGenerator,
                       private val entityReferenceLogic: EntityReferenceLogic,
                       private val invoiceDAO: InvoiceDAO,
                       private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<Invoice, InvoiceDAO>(sessionLogic), InvoiceLogic {

    override suspend fun createInvoice(invoice: Invoice) = fix(invoice) { invoice ->
        invoiceDAO.create(invoice)
    }

    override suspend fun deleteInvoice(invoiceId: String): DocIdentifier? {
        return try {
            deleteEntities(listOf(invoiceId)).firstOrNull()
        } catch (e: Exception) {
            throw DeletionException(e.message, e)
        }
    }

    override suspend fun getInvoice(invoiceId: String): Invoice? {
        return invoiceDAO.get(invoiceId)
    }

    override fun getInvoices(ids: List<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.getEntities(ids))
    }

    override suspend fun modifyInvoice(invoice: Invoice) = fix(invoice) { invoice ->
        invoiceDAO.save(invoice)
    }

    override fun modifyInvoices(invoices: List<Invoice>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.save(invoices))
    }

    override suspend fun addDelegation(invoiceId: String, delegation: Delegation): Invoice? {
        val invoice = getInvoice(invoiceId)
        return delegation.delegatedTo?.let { healthcarePartyId ->
            invoice?.let { c ->
                invoiceDAO.save(c.copy(delegations = c.delegations + mapOf(
                        healthcarePartyId to setOf(delegation)
                )))
            }
        } ?: invoice
    }

    override suspend fun addDelegations(invoiceId: String, delegations: List<Delegation>): Invoice? {
        val invoice = getInvoice(invoiceId)
        return invoice?.let {
            return invoiceDAO.save(it.copy(
                    delegations = it.delegations +
                            delegations.mapNotNull { d -> d.delegatedTo?.let { delegateTo -> delegateTo to setOf(d) } }
            ))
        }
    }

    override fun findInvoicesByAuthor(hcPartyId: String, fromDate: Long?, toDate: Long?, paginationOffset: PaginationOffset<List<String>>): Flow<ViewQueryResultEvent> = flow {
        emitAll(invoiceDAO.findInvoicesByHcParty(hcPartyId, fromDate, toDate, paginationOffset.toComplexKeyPaginationOffset()))
    }

    override fun listInvoicesByHcPartyContacts(hcParty: String, contactIds: Set<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndContacts(hcParty, contactIds))
    }

    override fun listInvoicesByHcPartyAndRecipientIds(hcParty: String, recipientIds: Set<String?>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndRecipientIds(hcParty, recipientIds))
    }

    override fun listInvoicesByHcPartyAndPatientSks(hcParty: String, secretPatientKeys: Set<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndPatientFk(hcParty, secretPatientKeys))
    }

    override fun listInvoicesByHcPartySentMediumTypeInvoiceTypeSentDate(hcParty: String, sentMediumType: MediumType, invoiceType: InvoiceType, sent: Boolean, fromDate: Long?, toDate: Long?): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartySentMediumTypeInvoiceTypeSentDate(hcParty, sentMediumType, invoiceType, sent, fromDate, toDate))
    }

    override fun listInvoicesByHcPartySendingModeStatus(hcParty: String, sendingMode: String?, status: String?, fromDate: Long?, toDate: Long?): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartySendingModeStatus(hcParty, sendingMode, status, fromDate, toDate))
    }

    override fun listInvoicesByHcPartyAndGroupId(hcParty: String, inputGroupId: String): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndGroupId(inputGroupId, hcParty))
    }

    override fun listInvoicesByHcPartyAndRecipientIdsUnsent(hcParty: String, recipientIds: Set<String?>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndRecipientIdsUnsent(hcParty, recipientIds))
    }

    override fun listInvoicesByHcPartyAndPatientSksUnsent(hcParty: String, secretPatientKeys: Set<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByHcPartyAndPatientFkUnsent(hcParty, secretPatientKeys))
    }

    override fun listInvoicesByServiceIds(serviceIds: Set<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesByServiceIds(serviceIds))
    }

    override suspend fun mergeInvoices(hcParty: String, invoices: List<Invoice>, destination: Invoice?): Invoice? {
        if (destination == null) return null
        for (i in invoices) {
            deleteInvoice(i.id)
        }
        return modifyInvoice(destination.copy(invoicingCodes = destination.invoicingCodes + invoices.flatMap { it.invoicingCodes }))
    }

    override suspend fun validateInvoice(hcParty: String, invoice: Invoice?, refScheme: String, forcedValue: String?): Invoice? {
        if (invoice == null) return null
        var refScheme: String = refScheme

        return modifyInvoice(invoice.copy(
                sentDate = System.currentTimeMillis(),
                invoiceReference = if (forcedValue != null || !Strings.isNullOrEmpty(invoice.invoiceReference)) {
                    forcedValue
                } else {
                    if (refScheme == null) {
                        refScheme = "yyyy00000"
                    }
                    val ldt = invoice.invoiceDate?.let { FuzzyValues.getDateTime(it) }
                            ?: LocalDateTime.now(ZoneId.systemDefault())
                    val f: NumberFormat = DecimalFormat("00")
                    val startScheme = refScheme.replace("yyyy".toRegex(), "" + ldt.year).replace("MM".toRegex(), f.format(ldt.monthValue.toLong())).replace("dd".toRegex(), "" + f.format(ldt.dayOfMonth.toLong()))
                    val endScheme = refScheme.replace("0".toRegex(), "9").replace("yyyy".toRegex(), "" + ldt.year).replace("MM".toRegex(), "" + f.format(ldt.monthValue.toLong())).replace("dd".toRegex(), "" + f.format(ldt.dayOfMonth.toLong()))
                    val prefix = "invoice:" + invoice.author + ":xxx:"
                    val fix = startScheme.replace("0+$".toRegex(), "")
                    val reference = entityReferenceLogic.getLatest(prefix + fix)
                    if (reference == null || !reference.id.startsWith(prefix)) {
                        val prevInvoices = invoiceDAO.listInvoicesByHcPartyAndReferences(hcParty, endScheme, null, true, 1)
                        val first = prevInvoices.firstOrNull()
                        "" + if (first?.invoiceReference != null) max(java.lang.Long.valueOf(first.invoiceReference) + 1L, java.lang.Long.valueOf(startScheme) + 1L) else java.lang.Long.valueOf(startScheme) + 1L
                    } else {
                        fix + (reference.id.substring(prefix.length + fix.length).toInt() + 1)
                    }
                }
        ))
    }

    override fun appendCodes(hcPartyId: String, userId: String, insuranceId: String?, secretPatientKeys: Set<String>, type: InvoiceType, sentMediumType: MediumType, invoicingCodes: List<InvoicingCode>, invoiceId: String?, invoiceGraceTime: Int?): Flow<Invoice> = flow {
        val fixedCodes = if (sentMediumType == MediumType.efact) {
            invoicingCodes.map { c -> c.copy(pending = true) }
        } else invoicingCodes
        val invoiceGraceTimeInDays = invoiceGraceTime ?: 0
        val selectedInvoice = if (invoiceId != null) getInvoice(invoiceId) else null
        var invoices = if (selectedInvoice != null) mutableListOf<Invoice>() else listInvoicesByHcPartyAndPatientSksUnsent(hcPartyId, secretPatientKeys)
                .filter { i -> i.invoiceType == type && i.sentMediumType == sentMediumType && if (insuranceId == null) i.recipientId == null else insuranceId == i.recipientId }.toList().toMutableList()
        if (selectedInvoice == null && invoices.isEmpty()) {
            invoices = listInvoicesByHcPartyAndRecipientIdsUnsent(hcPartyId, insuranceId?.let { setOf(it) }
                    ?: setOf()).filter { i -> i.invoiceType == type && i.sentMediumType == sentMediumType && i.secretForeignKeys == secretPatientKeys }.toList().toMutableList()
        }

        val modifiedInvoices: MutableList<Invoice> = LinkedList()
        val createdInvoices: MutableList<Invoice> = LinkedList()

        for (invoicingCode in fixedCodes) {
            val icDateTime = invoicingCode.dateCode?.let { FuzzyValues.getDateTime(it) }
            val unsentInvoice =
                    selectedInvoice
                            ?: if (icDateTime != null) invoices.filter { i ->
                                val invoiceDate = i.invoiceDate?.let { FuzzyValues.getDateTime(it) }
                                invoiceDate != null && abs(invoiceDate.withHour(0).withMinute(0).withSecond(0).withNano(0).until(icDateTime, ChronoUnit.DAYS)) <= invoiceGraceTimeInDays
                            }.firstOrNull()
                            else null

            if (unsentInvoice != null) {
                if (!createdInvoices.contains(unsentInvoice)) {
                    modifyInvoice(unsentInvoice.copy(
                            invoicingCodes = unsentInvoice.invoicingCodes + listOf(invoicingCode)
                    ))?.let {
                        modifiedInvoices.add(it)
                        emit(it)
                    }
                }
            } else {
                val now = System.currentTimeMillis()
                val newInvoice = Invoice(
                        id = uuidGenerator.newGUID().toString(),
                        invoiceDate = invoicingCode.dateCode ?: now,
                        invoiceType = type,
                        sentMediumType = sentMediumType,
                        recipientId = insuranceId,
                        recipientType = if (type == InvoiceType.mutualfund || type == InvoiceType.payingagency) Insurance::class.java.name else Patient::class.java.name,
                        invoicingCodes = listOf(invoicingCode),
                        author = userId,
                        responsible = hcPartyId,
                        created = now,
                        modified = now,
                        careProviderType = "persphysician",
                        invoicePeriod = 0,
                        thirdPartyPaymentJustification = "0"
                )

                //The invoice must be completed with ids and delegations and created on the server
                createdInvoices.add(newInvoice)
                emit(newInvoice)
                invoices.add(newInvoice)
            }
        }
    }

    override fun removeCodes(userId: String, secretPatientKeys: Set<String>, serviceId: String, inputTarificationIds: List<String>): Flow<Invoice> = flow {
        val tarificationIds = inputTarificationIds.toMutableList()
        val user = userLogic.getUser(userId)
        if (user != null) {
            val invoices = listInvoicesByHcPartyAndPatientSksUnsent(user.healthcarePartyId ?: throw IllegalArgumentException("The provided user must be linked to an hcp"), secretPatientKeys)
                    .filter { i -> i.invoicingCodes.any { ic -> serviceId == ic.serviceId && tarificationIds.contains(ic.tarificationId) } }
                    .toList().sortedWith(Comparator { a: Invoice, b: Invoice -> ((b.invoiceDate ?: 99999999999999L) as Long).compareTo(a.invoiceDate ?: 0L) })
            for (i in invoices) {
                var hasChanged = false
                val l: MutableList<InvoicingCode> = LinkedList(i.invoicingCodes)
                for (ic in i.invoicingCodes) {
                    if (tarificationIds.contains(ic.tarificationId)) {
                        l.remove(ic)
                        tarificationIds.remove(ic.tarificationId)
                        hasChanged = true
                    }
                }
                if (hasChanged) {
                    modifyInvoice(i.copy(invoicingCodes = l))?.let { emit(it) }
                }
            }
        }
    }

    override fun listInvoicesHcpsByStatus(status: String, from: Long?, to: Long?, hcpIds: List<String>): Flow<Invoice> = flow {
        emitAll(invoiceDAO.listInvoicesHcpsByStatus(status, from, to, hcpIds))
    }

    override fun solveConflicts(): Flow<Invoice> =
            invoiceDAO.listConflicts().mapNotNull { invoiceDAO.get(it.id, Option.CONFLICTS)?.let { invoice ->
                invoice.conflicts?.mapNotNull { conflictingRevision -> invoiceDAO.get(invoice.id, conflictingRevision) }
                        ?.fold(invoice) { kept, conflict -> kept.merge(conflict).also { invoiceDAO.purge(conflict) } }
                        ?.let { mergedInvoice -> invoiceDAO.save(mergedInvoice) }
            } }

    override suspend fun getTarificationsCodesOccurences(hcPartyId: String, minOccurences: Long): List<LabelledOccurence> {
        return invoiceDAO.listTarificationsFrequencies(hcPartyId)
                .filter { v -> v.value != null && v.value!! >= minOccurences }
                .map { v -> LabelledOccurence(v.key!!.components[1] as String, v.value) }
                .toList().sortedByDescending { it.occurence }
    }

    override fun listInvoicesIdsByTarificationsByCode(hcPartyId: String, codeCode: String, startValueDate: Long, endValueDate: Long): Flow<String> = flow {
        emitAll(invoiceDAO.listInvoiceIdsByTarificationsAndCode(hcPartyId, codeCode, startValueDate, endValueDate))
    }

    override fun listInvoiceIdsByTarificationsByCode(hcPartyId: String, codeCode: String?, startValueDate: Long?, endValueDate: Long?): Flow<String> = flow {
        emitAll(invoiceDAO.listInvoiceIdsByTarificationsByCode(hcPartyId, codeCode, startValueDate, endValueDate))
    }

    override fun filter(filter: FilterChain<Invoice>) = flow<Invoice> {
        val ids = filters.resolve(filter.filter).toList()
        val invoices = getInvoices(ids)
        val predicate = filter.predicate
        emitAll(if (predicate != null) invoices.filter { predicate.apply(it) } else invoices)
    }

    override fun getGenericDAO(): InvoiceDAO {
        return invoiceDAO
    }
}
